solr/src/main/java/org/apache/metamodel/solr/SolrDataContext.java

```diff
@@ -0,0 +1,403 @@
+                    .getEntity().getContent());
+
+            ObjectReader reader = new ObjectMapper().reader(HashMap.class);
+            Map<String, Object> outerMap = reader.readValue(inputStream);
+
+            Map<String, Object> schemaMap = null;
+
+            if (outerMap != null) {
+                schemaMap = (Map) outerMap.get("schema");
+            }
+
+            if (schemaMap != null) {
+                List<Map<String, String>> fieldMapList = (List) schemaMap
+                        .get("fields");
+                List<String> columnNames = new ArrayList<String>();
+                List<ColumnType> columnTypes = new ArrayList<ColumnType>();
+
+                for (int i = 0; i < fieldMapList.size(); i++) {
+                    Map<String, String> fieldMap = fieldMapList.get(i);
+
+                    columnNames.add(fieldMap.get("name"));
+                    columnTypes.add(ColumnType.STRING);
+                }
+
+                String[] columnNamesArr = new String[columnNames.size()];
+                columnNamesArr = columnNames.toArray(columnNamesArr);
+
+                ColumnType[] columnTypesArr = new ColumnType[columnTypes.size()];
+                columnTypesArr = columnTypes.toArray(columnTypesArr);
+
+                tableDef = new SimpleTableDef(indexName, columnNamesArr,
+                        columnTypesArr);
+            }
+        } catch (IOException e) {
+            logger.error("Failed to parse schema", "", e);
+        }
+
+        return tableDef;
+    }
+
+    @Override
+    protected Schema getMainSchema() throws MetaModelException {
+        try {
+            final MutableSchema theSchema = new MutableSchema(
+                    getMainSchemaName());
+
+            final MutableTable table = tableDef.toTable().setSchema(theSchema);
+            final Column idColumn = table.getColumnByName(FIELD_ID);
+            if (idColumn != null && idColumn instanceof MutableColumn) {
+                final MutableColumn mutableColumn = (MutableColumn) idColumn;
+                mutableColumn.setPrimaryKey(true);
+            }
+            theSchema.addTable(table);
+
+            return theSchema;
+        } catch (Exception e) {
+            throw new MetaModelException("Schema retrieval failed " + e);
+        }
+    }
+
+    @Override
+    protected Number executeCountQuery(Table table,
+            List<FilterItem> whereItems, boolean functionApproximationAllowed) {
+        String queryStr = "*:*";
+
+        if (whereItems != null) {
+            for (int i = 0; i < whereItems.size(); i++) {
+                FilterItem whereItem = whereItems.get(i);
+                queryStr = whereItem.toString();
+            }
+        }
+
+        QueryResponse response = selectRows(table, null, queryStr, 0, 0);
+        return response.getResults().getNumFound();
+    }
+
+    @Override
+    public DataSet executeQuery(Query q) throws MetaModelException {
```

Comment:
If I do not override executeQuery(...), then the Solr index cannot be queried by SQL, right? Also, I am calling super.executeQuery(q) at the end, to handle any other kind of query beyond the specific clauses I am handling. If the query is totally invalid and goes beyond any query which can be handled, exception may be thrown or DataSet null returned.

If you feel this approach still has some caveats, what method should be overridden so that the SQL interface still works?

What's your take?